### PR TITLE
sort: Fast path for already sorted data

### PR DESCRIPTION
**Pull request by bluss**
sort: Fast path for already sorted data

When merging two sorted blocks `left` and `right` if the last element in
`left` is <= the first in `right`, the blocks are already in sorted order.

Add this as an additional fast path by simply copying the whole left
block into the output and advancing the left pointer. The right block is
then treated the same way by the already present logic in the merge
loop.

Can reduce runtime of .sort() to less than 50% of the previous, if the data
was already perfectly sorted. Sorted data with a few swaps are also
sorted quicker than before. The overhead of one comparison per merge
seems to be negligible.

### CODE REVIEW

**Comment by bluss** [comment]
Out of tree benchmarks look like this:

“t” is for the number of elements in a random i32 vector,

`libstdsort` is a plain clone + sort of the random vector, the sorted variants are of the same vector but presorted.

I'm comparing my nightly rustc vs the patched rustc built from the tree. This results in a strange effect for the 10, 100 element case in the “after” build: what's happening is that an unoptimized)(?) jemalloc overhead dominates. Looks like a nice improvement when taking that into account.

```
before:

test t10000::libstd_sorted          ... bench:     429,275 ns/iter (+/- 1,836)
test t10000::libstd_sorted_10swaps  ... bench:     511,796 ns/iter (+/- 62,528)
test t10000::libstd_sorted_100swaps ... bench:     606,273 ns/iter (+/- 943)
test t10000::libstdsort             ... bench:     996,085 ns/iter (+/- 1,115)
test t1000::libstd_sorted           ... bench:      26,033 ns/iter (+/- 765)
test t1000::libstd_sorted_10swaps   ... bench:      35,167 ns/iter (+/- 507)
test t1000::libstd_sorted_100swaps  ... bench:      45,044 ns/iter (+/- 206)
test t1000::libstdsort              ... bench:      65,519 ns/iter (+/- 221)
test t100::libstd_sorted            ... bench:       1,842 ns/iter (+/- 9)
test t100::libstd_sorted_10swaps    ... bench:       2,675 ns/iter (+/- 20)
test t100::libstdsort               ... bench:       3,999 ns/iter (+/- 24)
test t10::libstd_sorted             ... bench:          74 ns/iter (+/- 1)
test t10::libstdsort                ... bench:         132 ns/iter (+/- 1)

after:

test t10000::libstd_sorted          ... bench:     175,627 ns/iter (+/- 747)
test t10000::libstd_sorted_10swaps  ... bench:     406,447 ns/iter (+/- 803)
test t10000::libstd_sorted_100swaps ... bench:     587,881 ns/iter (+/- 885)
test t10000::libstdsort             ... bench:     984,554 ns/iter (+/- 1,908)
test t1000::libstd_sorted           ... bench:      13,669 ns/iter (+/- 10)
test t1000::libstd_sorted_10swaps   ... bench:      34,950 ns/iter (+/- 90)
test t1000::libstd_sorted_100swaps  ... bench:      46,550 ns/iter (+/- 113)
test t1000::libstdsort              ... bench:      66,939 ns/iter (+/- 283)
test t100::libstd_sorted            ... bench:       5,450 ns/iter (+/- 10)
test t100::libstd_sorted_10swaps    ... bench:       6,798 ns/iter (+/- 9)
test t100::libstdsort               ... bench:       8,114 ns/iter (+/- 16)
test t10::libstd_sorted             ... bench:       2,204 ns/iter (+/- 4)
test t10::libstdsort                ... bench:       2,266 ns/iter (+/- 3)
```

If someone wants to run the in-tree benchmarks, I'd be thrilled to see that.

File: src/libcollections/slice.rs [hunk 1]
```diff
@@ -1066,6 +1066,16 @@ fn merge_sort<T, F>(v: &mut [T], mut compare: F) where F: FnMut(&T, &T) -> Order
                 let mut out = buf_tmp.offset(start as isize);
                 let out_end = buf_tmp.offset(right_end_idx as isize);
 
+                // if left[last] <= right[0], they are already in order:
+                // fast-forward the left side (the right side is handled
+                // in the loop).
+                if compare(&*right.offset(-1), &*right) != Greater {
```
Comment (bluss) [inline src/libcollections/slice.rs]: This needs a bounds check actually.

**Comment by Gankro** [comment]
Great wins on big runs! :+1: 

I'm a bit confused with what's happening in the small input cases. You mention jemalloc but I don't follow why that would have changed between runs?

**Comment by bluss** [comment]
> I'm comparing my nightly rustc vs the patched rustc built from the tree.

It's this.

**Comment by huonw** [comment]
>> I'm comparing my nightly rustc vs the patched rustc built from the tree.

> It's this.

I don't understand why that's relevant? If optimisations are enabled, they should be generating essentially the same code?

**Comment by huonw** [comment]
Oh, I see, if the patched rustc was built without optimisations, then jemalloc may've been too, and hence runs slower... This is somewhat suspicious, so r=me, except for nailing down the exact cause of the regression (and/or getting a version of the benchmarks that shows it doesn't exist for a "real" build).

**Comment by bluss** [comment]
Ok, I ran the in tree benchmarks:

```
before

test slice::bench::sort_big_random_large                        ... bench:   4,749,545 ns/iter (+/- 948,440) = 67 MB/s
test slice::bench::sort_big_random_medium                       ... bench:      36,870 ns/iter (+/- 812) = 86 MB/s
test slice::bench::sort_big_random_small                        ... bench:       1,798 ns/iter (+/- 100) = 88 MB/s
test slice::bench::sort_big_sorted                              ... bench:     772,643 ns/iter (+/- 11,515) = 414 MB/s
test slice::bench::sort_random_large                            ... bench:   1,806,359 ns/iter (+/- 36,855) = 44 MB/s
test slice::bench::sort_random_medium                           ... bench:      14,509 ns/iter (+/- 198) = 55 MB/s
test slice::bench::sort_random_small                            ... bench:         685 ns/iter (+/- 20) = 58 MB/s
test slice::bench::sort_sorted                                  ... bench:     397,642 ns/iter (+/- 1,082) = 100 MB/s

after

test slice::bench::sort_big_random_large                        ... bench:   4,718,159 ns/iter (+/- 41,542) = 67 MB/s
test slice::bench::sort_big_random_medium                       ... bench:      36,909 ns/iter (+/- 292) = 86 MB/s
test slice::bench::sort_big_random_small                        ... bench:       1,888 ns/iter (+/- 13) = 84 MB/s
test slice::bench::sort_big_sorted                              ... bench:     615,455 ns/iter (+/- 8,196) = 519 MB/s
test slice::bench::sort_random_large                            ... bench:   1,796,680 ns/iter (+/- 13,319) = 44 MB/s
test slice::bench::sort_random_medium                           ... bench:      14,588 ns/iter (+/- 352) = 54 MB/s
test slice::bench::sort_random_small                            ... bench:         699 ns/iter (+/- 9) = 57 MB/s
test slice::bench::sort_sorted                                  ... bench:     134,979 ns/iter (+/- 1,101) = 296 MB/s
```

Which I think looks good?

**Comment by bluss** [comment]
The "big" cases sort elements of `type BigSortable = (u64, u64, u64, u64); `, small cases u64, which makes a surprising difference.

**Comment by Gankro** [comment]
@bluss thanks for following up! :+1:

**Pull request merged**